tutorials/identifying-first-day-no-snow/index.md

```diff
@@ -0,0 +1,516 @@
+
+```js
+var completeCol = ee.ImageCollection('MODIS/006/MOD10A1')
+  .select('NDSI_Snow_Cover');
+```
+
+Mask pixels based on frequency of snow cover.
+
+```js
+// Pixels must have been 10% snow covered for at least 2 weeks in 2018.
+var snowCoverEphem = completeCol.filterDate('2018-01-01', '2019-01-01')
+  .map(function(img) {
+    return img.gte(10);
+  })
+  .sum()
+  .gte(14);
+
+// Pixels must not be 10% snow covered more than 124 days in 2018.
+var snowCoverConst = completeCol.filterDate('2018-01-01', '2019-01-01')
+  .map(function(img) {
+    return img.gte(10);
+  })
+  .sum()
+  .lt(125);
+```
+
+Combine the water mask and the snow cover frequency masks.
+
+```js
+var analysisMask = waterMask.multiply(snowCoverEphem).multiply(snowCoverConst);
+```
+
+### 4\. Identify the first day of the year without snow per pixel, per year
+
+Make a list of the years to process; input variables were defined in step 1.
+
+```js
+var years = ee.List.sequence(startYear, endYear);
+```
+
+Map the following function over the list of years. For each year, identify
+the first day with zero percent snow cover.
+ 
+1. Define the start and end dates to filter the dataset for the given year.
+2. Filter the image collection by the date range.
+3. Add the date bands to each image in the filtered collection.
+4. Sort the filtered collection by date.
+5. Make a mosaic using the `min` reducer to select the pixel with 0 (minimum)
+snow cover. Since the collection is sorted by date, the first image with 0
+snow cover is selected. This operation is conducted per-pixel to build the
+complete image mosaic.
+6. Apply the analysis mask to the resulting mosaic.
+ 
+An `ee.List` of images is returned.
+
+```js
+var annualList = years.map(function(year) {
+  // Set the global startYear variable as the year being worked on so that
+  // it will be accessible to the addDateBands mapped to the collection below.
+  startYear = year;
+  // Get the first day-of-year for this year as an ee.Date object.
+  var firstDoy = ee.Date.fromYMD(year, 1, 1);
+  // Advance from the firstDoy to the user-defined startDay; subtract 1 since
+  // firstDoy is already 1. Set the result as the global startDate variable so
+  // that it is accessible to the addDateBands mapped to the collection below.
+  startDate = firstDoy.advance(startDoy-1, 'day');
+  // Get endDate for this year by advancing 1 year from startDate.
+  // Need to advance an extra day because end date of filterDate() function
+  // is exclusive.
+  var endDate = startDate.advance(1, 'year').advance(1, 'day');
+  // Filter the complete collection by the start and end dates just defined.
+  var yearCol = completeCol.filterDate(startDate, endDate);
+  // Construct an image where pixels represent the first day within the date range
+  // that the lowest snow fraction is observed.
+  var noSnowImg = yearCol
+    // Add date bands to all images in this particular collection.
+    .map(addDateBands)
```

Comment:
In that case `addDateBands` here would be replaced with `function(img) { return addDateBands(img, startYear, startDate); }`, for example.

Reply:
Altering the `addDateBands` function to include `startYear` and `startDate` parameters and applying it in an anonymous mapped function does not work because `startYear` and `startDate` are still outside the scope.

https://code.earthengine.google.com/7f8d90d5060b4272ef2a3244eb04d55b?as_external

I'm happy to try something else.

Reply:
Original reply somehow got lost: I was saying that you're absolutely correct; getting those variables to be shared across scopes would require more advanced JavaScript that is probably out of the scope of this tutorial, so please feel free to keep the globals!

Reply:
Ack